File: src/main/java/org/taktik/icure/dao/replicator/AbstractReplicator.java
package org.taktik.icure.dao.replicator;

import com.google.common.collect.ImmutableList;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IMap;
import kotlin.Unit;
import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.util.ssl.SslContextFactory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.taktik.couchdb.Change;
import org.taktik.couchdb.CouchDbInstance;
import org.taktik.icure.entities.Group;
import org.taktik.icure.entities.base.StoredDocument;

import javax.annotation.PostConstruct;
import java.net.URI;
import java.util.List;
import java.util.concurrent.CompletableFuture;

import static org.taktik.icure.dao.replicator.NewGroupObserver.IS_SYNC_DEAD_TIMEOUT;

/**
 * @author Bernard Paulus - 14/03/2017
 */
public abstract class AbstractReplicator<T extends StoredDocument> implements FilteredReplicator {
	private static final Logger log = LoggerFactory.getLogger(AbstractReplicator.class);

	private HazelcastInstance hazelcast;
	private IMap<String, ReplicatorJobStatus> replicatorJobsStatusesByGroupId;

	@Value("${icure.couchdb.username}")
	protected String couchDbUsername;
	@Value("${icure.couchdb.password}")
	protected String couchDbPassword;
	@Value("${icure.couchdb.prefix}")
	protected String couchDbPrefix;
	@Value("${icure.couchdb.url}")
	protected String couchDbUrl;

	protected HttpClient httpClient = null;

	public AbstractReplicator(HazelcastInstance hazelcast) {
		this.hazelcast = hazelcast;
	}

	protected abstract List<String> getAllIds(String groupId);

	protected abstract Class<T> getEntityType();

	protected abstract void prepareReplication(Group group);

	protected abstract void replicate(Group group, List<String> entityIds);

	@PostConstruct
	public void init() {
		replicatorJobsStatusesByGroupId = hazelcast.getMap(getClass().getCanonicalName() + ".replicatorJobsStatusesByGroupId");
	}

	@Override
	public CompletableFuture<Boolean> startReplication(Group replicatedGroup, SslContextFactory sslContextFactory) {
		// should check if a replication hasn't been started or if group has been updated
		GroupDBUrl groupDb = new GroupDBUrl(couchDbUrl);
		ReplicatorJobStatus jobStatus = tryToGetJob(replicatedGroup);
		if (jobStatus == null) {
			log.info("Someone else is busy with " + replicatedGroup.getId());
			return CompletableFuture.completedFuture(true); //It is a success because someone else is taking care of this replication... We do not need to mark it as pending
		}
		synchronized (this) {
			if (httpClient == null) {
				log.info("Init http client from " + replicatedGroup.getId());
				httpClient = new HttpClient(sslContextFactory);
				try {
					httpClient.setMaxConnectionsPerDestination(65535);
					httpClient.start();
				} catch (Exception e) {
					log.error("Cannot start HTTP client");
					try {
						httpClient.stop();
					} catch (Exception ignored) {
						return CompletableFuture.completedFuture(false); //retry
					} finally {
						httpClient = null;
					}
					return CompletableFuture.completedFuture(false); //retry
				}
			}
		}
		CouchDbInstance groupDbInstance = new CouchDbInstance(httpClient, URI.create(groupDb.getInstanceUrl(replicatedGroup)), groupDb.getDbName(replicatedGroup), replicatedGroup.getId(), replicatedGroup.getPassword());
		log.info("Create instance for " + replicatedGroup.getId());
		return groupDbInstance.exists().thenApply((Boolean result) -> {
			log.info("Db exists for " + replicatedGroup.getId());
			if (result) {
				prepareReplication(replicatedGroup);
				result = replicateExistingData(replicatedGroup); //This is sort of blocking
				if (result) {
					observeChanges(replicatedGroup, jobStatus, groupDbInstance);
				}
			}
			return result;
		});
	}

	private void observeChanges(Group replicatedGroup, ReplicatorJobStatus jobStatus, CouchDbInstance groupDbInstance) {
		groupDbInstance.changes(jobStatus.getSeq() != null ? jobStatus.getSeq() : "now", (Change c) -> {
			observe(c, replicatedGroup);
			return Unit.INSTANCE;
		}, () -> {
			replicatorJobsStatusesByGroupId.put(replicatedGroup.getId(), replicatorJobsStatusesByGroupId.get(replicatedGroup.getId()).timestamp(System.currentTimeMillis()));
			return Unit.INSTANCE;
		}).whenComplete((unit, throwable) -> observeChanges(replicatedGroup, jobStatus, groupDbInstance));
	}

	private boolean replicateExistingData(Group group) {
		long startTime = System.currentTimeMillis();
		try {
			List<String> allIds = this.getAllIds(group.getId());
			replicate(group, allIds);
		} catch (Exception e) {
			log.error("Exception during initial replication : {} for group {} ({}) ", e.getLocalizedMessage(), group.getId(), group.getName());
			return false;
		}
		long endTime = System.currentTimeMillis();
		long duration = (endTime - startTime);
		log.info("Initial replication completed for {} in {} ms", group.getId(), duration);
		return true;
	}

	public void observe(Change change, Group group) {
		if (change.getDoc() == null) {
			return;
		}
		if (getEntityType().getCanonicalName().equals(change.getDoc().get("java_type"))) {
			ImmutableList<String> entityIds = ImmutableList.of((String) change.getDoc().get("_id"));
			try {
				replicate(group, entityIds);
			} catch (Exception e) {
				log.error("replication failed: {}", e.getLocalizedMessage());
				return;
			}
			replicatorJobsStatusesByGroupId.put(group.getId(), new ReplicatorJobStatus(System.currentTimeMillis(), change.getSeq()));
		}
	}

	private ReplicatorJobStatus tryToGetJob(Group group) {
		final ReplicatorJobStatus jobStatus = new ReplicatorJobStatus(System.currentTimeMillis(), null);
		ReplicatorJobStatus jobStatusInCluster = replicatorJobsStatusesByGroupId.putIfAbsent(group.getId(), jobStatus);
		if (jobStatusInCluster == null) {
			return jobStatus;
		} else {
			if (jobStatus.getTimestamp() - jobStatusInCluster.getTimestamp() <= IS_SYNC_DEAD_TIMEOUT /* the current job is not dead */ || !replicatorJobsStatusesByGroupId.replace(group.getId(), jobStatusInCluster, jobStatus)) /* It has already been acquired in the meantime */ {
				return null;
			} else {
				return jobStatus.seq(jobStatusInCluster.getSeq());
			}
		}
	}

}

File: src/main/kotlin/org/taktik/icure/dao/replicator/AbstractReplicator.kt
package org.taktik.icure.dao.replicator

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.collect
import org.eclipse.jetty.client.HttpClient
import org.eclipse.jetty.util.ssl.SslContextFactory
import org.ektorp.http.URI
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.taktik.couchdb.Client
import org.taktik.couchdb.ClientImpl
import org.taktik.icure.entities.Group
import org.taktik.icure.entities.base.StoredDocument

abstract class AbstractReplicator<T : StoredDocument>(private val sslContextFactory: SslContextFactory) : Replicator {
    private val log = LoggerFactory.getLogger(AbstractReplicator::class.java)

    @Value("\${icure.couchdb.username}")
    protected var couchDbUsername: String? = null
    @Value("\${icure.couchdb.password}")
    protected var couchDbPassword: String? = null
    @Value("\${icure.couchdb.prefix}")
    protected var couchDbPrefix: String? = null
    @Value("\${icure.couchdb.url}")
    protected lateinit var couchDbUrl: String

    protected val httpClient: HttpClient by lazy {
        HttpClient(this.sslContextFactory).apply {
            try {
                maxConnectionsPerDestination = 65535
                maxRequestsQueuedPerDestination = 4096
                start()
            } catch (e: Exception) {
                log.error("Cannot start HTTP client", e)
                try {
                    stop()
                } catch (ignored: Exception) {
                }
                throw e
            }
        }
    }

    protected abstract val entityType: Class<T>

    protected abstract suspend fun getAllIds(groupId: String, dbInstanceUrl: String?): List<String>

    protected abstract suspend fun prepareReplication(group: Group)

    protected abstract suspend fun replicate(group: Group, entityIds: List<String>): List<String>

    @FlowPreview
    @ExperimentalCoroutinesApi
    override suspend fun startReplication(group: Group): Job {
        val groupDb = GroupDBUrl(couchDbUrl)
        val dbURI = URI.of(groupDb.getInstanceUrl(group)).append(groupDb.getDbName(group))
        val client = ClientImpl(httpClient, dbURI, group.id, group.password)

        require(client.exists()) { "Cannot start replication : the group db doesnt exist for ${group.id}" }
        log.info("Db exists for ${group.id}")
        prepareReplication(group)
        replicateExistingData(group)
        // FIXME We should replicateExistingData every X to ensure we didn't miss any change
        return GlobalScope.launch {
            observeChanges(group, client)
        }
    }

    @ExperimentalCoroutinesApi
    @FlowPreview
    private suspend fun observeChanges(group: Group, dbClient: Client) {
        val changes = dbClient.subscribeForChanges(entityType)
        // Replicate
        changes.collect { change ->
            log.debug("Detected new object : ${change.id} in group ${group.id}")
            val entityIds = listOf(change.doc.id)
            replicate(group, entityIds)
        }
    }

    private suspend fun replicateExistingData(group: Group): Boolean {
        val startTime = System.currentTimeMillis()
        try {
            val allIds = this.getAllIds(group.id, group.dbInstanceUrl())
            log.debug("Replicating ${allIds.size} existing objects in group ${group.id}")
            replicate(group, allIds)
            log.debug("Done replicating ${allIds.size} existing objects in group ${group.id}")
        } catch (e: Exception) {
            log.error("Exception during initial replication : ${e.localizedMessage} for group ${group.id} (${group.name})", e)
            return false
        }

        val endTime = System.currentTimeMillis()
        val duration = endTime - startTime
        log.info("Initial replication completed for group ${group.id} in $duration ms")
        return true
    }
}